src/html5/parser/document.rs

```diff
@@ -60,6 +62,153 @@ impl DocumentFragment {
+    CreateElement {
+        name: String,
+        parent_id: NodeId,
+        position: Option<usize>,
+        namespace: String,
+    },
+    CreateText {
+        content: String,
+        parent_id: NodeId,
+    },
+    CreateComment {
+        content: String,
+        parent_id: NodeId,
+    },
+    InsertAttribute {
+        key: String,
+        value: String,
+        element_id: NodeId,
+    },
+}
+
+/// Queue of tasks that will mutate the document to add/update
+/// nodes in the tree. These tasks are performed sequentially in the
+/// order they were created by the TreeBuilder.
+pub struct DocumentTaskQueue {
+    /// Internal counter of the next ID to generate from the NodeArena
+    /// without actually registering the node.
+    /// WARNING: if nodes are registered in the arena while tasks are being queued
+    /// this could lead to conflicts in NodeIds. NodeArena should NOT be used directly
+    /// if using a DocumentTaskQueue.
+    next_node_id: NodeId,
+    /// Reference to the document to commit changes to
+    pub(crate) document: DocumentHandle,
+    /// List of tasks to commit upon execute() which is cleared after execution finishes.
+    // IMPLEMENTATION NOTE: using a vec here since I'm assuming we are
+    // executing all tasks at once. If we need to support stopping task
+    // execution midway, then maybe a real "queue" structure that pops
+    // completed tasks is needed.
+    pub(crate) tasks: Vec<DocumentTask>,
+}
+
+// See tree_builder.rs for method comments
+impl TreeBuilder for DocumentTaskQueue {
+    fn is_empty(&self) -> bool {
+        self.tasks.is_empty()
+    }
+
+    fn create_element(
+        &mut self,
+        name: &str,
+        parent_id: NodeId,
+        position: Option<usize>,
+        namespace: &str,
+    ) -> NodeId {
+        let element = DocumentTask::CreateElement {
+            name: name.to_owned(),
+            parent_id,
+            position,
+            namespace: namespace.to_owned(),
+        };
+        let new_id = self.next_node_id;
+        self.next_node_id = self.next_node_id.next();
+        self.tasks.push(element);
+
+        new_id
+    }
+
+    fn create_text(&mut self, content: &str, parent_id: NodeId) {
+        let text = DocumentTask::CreateText {
+            content: content.to_owned(),
+            parent_id,
+        };
+        self.tasks.push(text);
+    }
+
+    fn create_comment(&mut self, content: &str, parent_id: NodeId) {
+        let comment = DocumentTask::CreateComment {
+            content: content.to_owned(),
+            parent_id,
+        };
+        self.tasks.push(comment);
+    }
+
+    fn insert_attribute(&mut self, key: &str, value: &str, element_id: NodeId) {
+        let attribute = DocumentTask::InsertAttribute {
+            key: key.to_owned(),
+            value: value.to_owned(),
+            element_id,
+        };
+        self.tasks.push(attribute);
+    }
+
+    fn execute(&mut self) -> Vec<String> {
```

Comment:
I think "flush" is the term I've been seeing at the level of the queue of tasks.

Reply:
I was swapping back and forth between `flush` and `execute`, I don't really have a strong opinion so I'm open to renaming